### ISSUE TITLE
Regression in memory allocation optimization of a mutable StaticArray

### ISSUE THREAD

**Issue by cdsousa** [JuliaLang/julia#41800]
Julia version 1.6.2 was able to optimize out the memory allocation of the following code, but version1.7.0-beta3.0 is not:
```julia
using StaticArrays
function cpu_kernel(a, i)
    SIZE = 15
    v = ones(MVector{SIZE, UInt32})
    for x in v
        a[i] += x
    end
end
a = zeros(UInt32, 1)
cpu_kernel(a, 1)
@time cpu_kernel(a, 1)
```
This has been noticed originally here: https://github.com/JuliaGPU/CUDA.jl/issues/38#issuecomment-894315291

(Maybe related to https://github.com/JuliaLang/julia/issues/41512 ?)

**Comment by cdsousa** [comment]
This also allocates in Version 1.8.0-DEV.310 (2021-08-06).

**Comment by KristofferC** [comment]
A bisection might be beneficial here to find the offending commit.

**Comment by maleadt** [comment]
I'm bisecting this.

**Comment by maleadt** [comment]
Reduced Julia example:

```julia
import Base: getindex, ones
struct SOneTo{a} <: AbstractUnitRange{Int}
end
SOneTo(a) = SOneTo{a}()
Base.first(::SOneTo) = 1
Base.last(::SOneTo{a}) where a = a
function Base.iterate(::SOneTo{a}, b) where a
    if b < a
        c = b + 1
        c, c
    end
end
abstract type StaticArray{a, c, b} <: AbstractArray{c, b} end
struct Size{a}
    end
Size(b) = Size{tuple(b.parameters...)}()
Size(::c) where c<: StaticArray{a} where a= @isdefined(a) ? Size(a) : d
mutable struct MArray{a , b, c, d} <: StaticArray{a, b, c}
    function MArray{a,b,c,d}(e) where {a,b,c,d}
        new()
    end
end
function getindex(f::MArray, h)
    checkbounds(f,h)
      g
end
MVector{a, b} = MArray{Tuple{a}, b, 1, a}
Base.axes(a::StaticArray) = b(Size(a))
function b(::Size{c}) where c
    map(SOneTo, c) end
ones(::Type{a}) where a = b(Size, a)
@generated function b(::c, ::Type{a}) where {c, a }
    quote
        a(0)
    end
end
function b(a, c)
    d = 15
    e = ones(MVector{d, UInt32})
    for f in e
    end
end
a = using InteractiveUtils
@code_llvm b(a, 1)
```

And the IR that fails to optimize:

```llvm
target datalayout = "f80:128-n64-S128"

@a = external global {}*

define void @b() {
  %c = call {} addrspace(10)* @julia.gc_alloc_obj(i8* undef, i64 0, {} addrspace(10)* null)
  br label %d

ar:
  br label %ar.as

ar.as:
  br label %e

e:
  %at = phi i64 [ 1, %ar.as ], [ %f, %g ]
  %aa = icmp slt i64 %at, 15
  %h = zext i1 %aa to i8
  %i = trunc i8 %h to i1
  %j = xor i1 %i, true
  br i1 %j, label %k, label %l

l:
  %ab = add i64 %at, 1
  br label %m

k:
  br label %m

m:
  %n = phi i8 [ 0, %l ], [ 1, %k ]
  %ac = phi i64 [ %ab, %l ], [ undef, %k ]
  %o = phi i64 [ %ab, %l ], [ undef, %k ]
  %p = trunc i8 %n to i1
  %ad = xor i1 %p, true
  br i1 %ad, label %q, label %r

r:
  br label %s

q:
  %au = icmp sle i64 1, %ac
  %t = icmp sle i64 %ac, 15
  %ae = zext i1 %au to i8
  %af = zext i1 %t to i8
  %ag = and i8 %ae, %af
  %ah = trunc i8 %ag to i1
  %u = zext i1 %ah to i8
  %ai = trunc i8 %u to i1
  %v = xor i1 %ai, true
  br i1 %v, label %w, label %x

w:
  %1 = call {} addrspace(10)* @aj({} addrspace(10)* %c)
  unreachable

x:
  %y = load atomic {}*, {}** @a unordered, align 8
  %z = icmp ne {}* %y, null
  br i1 %z, label %ak, label %al

av:
  br label %s

s:
  %f = phi i64 [ %o, %av ], [ undef, %r ]
  %am = phi i8 [ 1, %r ], [ 0, %av ]
  %an = xor i8 %am, 1
  %aw = trunc i8 %an to i1
  %ao = xor i1 %aw, true
  br i1 %ao, label %ap, label %g

g:
  br label %e

ap:
  ret void

d:
  br label %ar

al:
  %2 = call {}* @aq()
  br label %ak

ak:
  br label %av
}

declare {} addrspace(10)* @julia.gc_alloc_obj(i8*, i64, {} addrspace(10)*)

declare {}* @aq()

declare {} addrspace(10)* @aj({} addrspace(10)*)
```

Contains an allocation after optimization on LLVM 12, not on LLVM 11. Can be checked with LLVM.jl on 1.6 and 1.7 using:

```julia
using LLVM

Context() do ctx
    ir = read(first(ARGS), String)
    mod = parse(LLVM.Module, ir; ctx)

    ModulePassManager() do pm
        ccall(:jl_add_optimization_passes, Cvoid,
                (LLVM.API.LLVMPassManagerRef, Cint, Cint),
                pm, Base.JLOptions().opt_level, #=lower_intrinsics=# 1)
        run!(pm, mod)
    end

    println(mod)
end
```

**Comment by vtjnash** [comment]
It is impressive that LLVM used to be able to remove the bounds check. We may want to start emitting explicit semantic copies on error paths though, to make that less challenging for it.

**Comment by JeffBezanson** [comment]
Seems to be due to the value escaping via a BoundsError, and LLVM no longer removing a bounds check. So we could work around it by adding a `copy` when throwing a BoundsError for this type. Obviously it would be great to fix this in LLVM, but that seems like it could be quite difficult.

**Comment by maleadt** [comment]
> We may want to start emitting explicit semantic copies on error paths though, to make that less challenging for it.

How would that look here with the MWE (which doesn't rely on StaticArrays.jl)? `@noinline Base.throw_boundserror(A::MArray, I) = throw(BoundsError(deepcopy(A), I))` doesn't seem to cut it.

**Comment by vtjnash** [comment]
`deepcopy` needs to escape the object (to detect reference cycles). Maybe a regular array copy?

**Comment by maleadt** [comment]
> Maybe a regular array copy?

Still allocates. `throw_boundserror` doesn't even need to capture the array, `@inline Base.throw_boundserror(f::MArray, I) = throw(I)` also causes an allocation.

**Comment by vtjnash** [comment]
true, I don't think we would optimize (or inline) that, because it is on the error path

**Comment by JeffBezanson** [comment]
So far this seems too difficult to fix to be release-blocking. We absolutely should fix it though, and can backport the fix to a future 1.7.x patch release.

**Comment by maleadt** [comment]
This seems fixed on 1.8, presumably because of https://github.com/JuliaLang/julia/pull/43057.
